### PR TITLE
Added swift mail spooler cleaner fixes #1

### LINKED ISSUE #1: Mail spooler cleaner
As suggested here: https://twitter.com/AllWrightyThen/status/575966358623682560

See also http://sgoettschkes.blogspot.de/2012/09/symfony-21-commands-and-swiftmailer.html

### PR DESCRIPTION
**Pull request by cmodijk**
Like the title it cleans the swift mailer spooler the code is based on the [EmailSenderListener](https://github.com/symfony/SwiftmailerBundle/blob/master/EventListener/EmailSenderListener.php) of the `SwiftmailerBundle`.

### CODE REVIEW

**Comment by matthiasnoback** [comment]
Thanks for creating this PR!

I've got two concerns about the code: 
1. Currently it's pretty much coupled to Symfony (because of the container dependency and the hard-coded service ids). I would like the LongRunning project to be useful in other applications as well, like Silex applications (see also http://silex.sensiolabs.org/doc/providers/swiftmailer.html). So I'd say: we'd have to use a compiler pass to collect all the transport/spool services that are relevant and inject those as constructor arguments into this particular cleaner.
2. The amount of nested if-statements is a bit too high to my taste ;) But I think this should be automatically fixed when we 1. has been fixed.

**Comment by cmodijk** [comment]
I'm currently working on implementing the compiler but i'm wondering how I should implement the instanceof a class inside a compiler pass. Should I get the service and do al the instanceof checks there?   Juist checking the class name from the definition is not enough because the `SpoolTransport` class extends the `Swift_Transport_SpoolTransport`.

**Comment by matthiasnoback** [comment]
You could take the class from the definition, then resolve its actual value using the `ParameterBag::resolveValue()` function. You should then be able to use the `is_a()` standard PHP function to do the check.

**Comment by cmodijk** [comment]
I'm completely stuck with this one this is what I have so fare https://gist.github.com/cmodijk/9ac56c22f34fc2185476 i'm getting a DefinitionDecorator and there is no class inside that is just empty.

**Comment by matthiasnoback** [comment]
Okay, right, then I'd leave it as it was using the `PassConfig::OPTIMIZE'
option.

There are no tests at all for this library. If you can think of something,
you're welcome to add a unit/integration test but if not, then it's just
fine!

Do you think this is ready to be merged as it is?
Op 10 apr. 2015 18:42 schreef "Cliff Odijk" notifications@github.com:

**Comment by cmodijk** [comment]
The `PassConfig::OPTIMIZE` part is already replaced inside the Monolog pull request #4 and that worked over there. I have not tried using the `PassConfig::OPTIMIZE` for the SwiftMailer pull request. Will try that a little later.

**Comment by matthiasnoback** [comment]
It should work (or maybe even use `REMOVE`) because at that compile stage
the definition decorators are being replaced by regular service definitions.
Op 10 apr. 2015 18:50 schreef "Cliff Odijk" notifications@github.com:

**Comment by cmodijk** [comment]
I just updated my pull request. I used the `TYPE_OPTIMIZE` because during the `TYPE_REMOVE` proces the `long_running.swift_mailer.clear_spools` service which is private is not there any more. I think this should be it but I would like to hear any feedback you have because your comment about nested if-statements is now just moved to the `CompilerPass`.

**Comment by matthiasnoback** [comment]
Looks good to me!

A small improvement: I'd prefer not to use `!` in `if` statements where the
value of the expression is not boolean. E.g. inside the compiler pass it
would be better compare with 'null`or`array`.

By the way since it's inside the compiler pass I don't mind the nested if
statements :)
Op 11 apr. 2015 23:16 schreef "Cliff Odijk" notifications@github.com:

**Comment by cmodijk** [comment]
Do you mean the if's on line [14](https://github.com/LongRunning/LongRunning/pull/3/files#diff-56014eb077fea8a2cce13a25ddc808d5R14) or [18](https://github.com/LongRunning/LongRunning/pull/3/files#diff-56014eb077fea8a2cce13a25ddc808d5R18) or both?

**Comment by matthiasnoback** [comment]
Only the one one line 18 - it should be `$container->findDefinition('swiftmailer.mailer') === null || $container->getParameter('swiftmailer.mailers') === []`. For the one on 14 already the return value of `has()` is already a boolean value.

**Comment by matthiasnoback** [comment]
By the way, would you like to add a few quick notes in the README file about the cleaners that you contributed?

**Comment by cmodijk** [comment]
Done! Please spel check the README for me.

File: README.md [hunk 1]
```diff
@@ -10,6 +10,9 @@ In a Symfony application enable the bundle `LongRunning\Bundle\LongRunningBundle
 - Clear all Doctrine ORM entity managers (to prevent outdated entities from being updated)
 - Reset all closed Doctrine ORM entity managers (after a failed transaction)
 - Close all database connections (to prevent database timeout errors)
+- Clear all monolog fingers crossed handlers (clears messages and resets if it had failed during the command)
```
Comment (matthiasnoback) [inline README.md]: What about:

> - Clear all Monolog "fingers crossed" handlers (clears messages and resets the handler when there was no failure during the execution of a task)
> - Close all Monolog buffer handlers (clears log messages that were buffered during the execution of a task)
> - Flush all Swift Mailer "in memory" spools (i.e. send spooled e-mails)

**Comment by matthiasnoback** [comment]
Thanks again, almost ready to merge! I really like this project, eventually this might be a home to many more libraries. I'm thinking about signal handling, and PID file creation, etc.

**Comment by cmodijk** [comment]
Tnx for al the feedback! I updated the pull request.

**Comment by matthiasnoback** [comment]
No, thank _you_! :)

**Pull request merged**